diff --git a/src/cairo-quartz-surface.c b/src/cairo-quartz-surface.c
index 7a9f524..45305dd 100644
--- a/src/cairo-quartz-surface.c	
+++ b/src/cairo-quartz-surface.c
@@ -794,7 +794,7 @@ _cairo_surface_to_cgimage (cairo_surface_t       *source,
 {
     cairo_status_t status;
     cairo_image_surface_t *image_surface;
-    void *image_data, *image_extra;
+    void *image_extra;
     cairo_bool_t acquired = FALSE;
 
     if (source->backend && source->backend->type == CAIRO_SURFACE_TYPE_QUARTZ_IMAGE) {
@@ -855,28 +855,15 @@ _cairo_surface_to_cgimage (cairo_surface_t       *source,
 	return status;
     }
 
-    image_data = _cairo_malloc_ab (image_surface->height, image_surface->stride);
-    if (unlikely (!image_data))
-    {
-	if (acquired)
-	    _cairo_surface_release_source_image (source, image_surface, image_extra);
-	else
-	    cairo_surface_destroy (&image_surface->base);
-
-	return _cairo_error (CAIRO_STATUS_NO_MEMORY);
-    }
-
-    memcpy (image_data, image_surface->data,
-	    image_surface->height * image_surface->stride);
     *image_out = CairoQuartzCreateCGImage (image_surface->format,
 					   image_surface->width,
 					   image_surface->height,
 					   image_surface->stride,
-					   image_data,
+					   image_surface->data,
 					   TRUE,
 					   NULL,
-					   DataProviderReleaseCallback,
-					   image_data);
+					   NULL,
+					   NULL);
 
     /* TODO: differentiate memory error and unsupported surface type */
     if (unlikely (*image_out == NULL))
