diff --git a/src/cairo-quartz-surface copy.c b/src/cairo-quartz-surface.c
index b8dc293..b572e28 100644
--- a/src/cairo-quartz-surface copy.c	
+++ b/src/cairo-quartz-surface.c
@@ -2673,7 +2673,7 @@ _cairo_quartz_image_to_png (CGImageRef image, const char *dest)
 
     memset (pathbuf, 0, sizeof (pathbuf));
     dest = dest ? dest : image_name;
-    snprintf (pathbuf, sizeof (pathbuf), "%s/Desktop/%s%d.png",getenv ("HOME"), dest, sctr++, ext);
+    snprintf (pathbuf, sizeof (pathbuf), "%s/Desktop/%s%d.png",getenv ("HOME"), dest, sctr++);
     path = CFStringCreateWithCString (NULL, pathbuf, kCFStringEncodingUTF8);
     url = CFURLCreateWithFileSystemPath (NULL, path, kCFURLPOSIXPathStyle, FALSE);
     image_dest = CGImageDestinationCreateWithURL (url, png_utti, 1, NULL);
