diff --git a/configure copy.ac b/configure.ac
index 2446551..fe644ad 100644
--- a/configure copy.ac	
+++ b/configure.ac
@@ -1264,7 +1264,7 @@ AC_ARG_ENABLE(quartz-relocation,
                               [quartz_relocation=yes])
 
 if test "x$gdktarget" = "xquartz"; then
-  GDK_EXTRA_LIBS="$GDK_EXTRA_LIBS -framework Cocoa"
+  GDK_EXTRA_LIBS="$GDK_EXTRA_LIBS -framework Cocoa -framework QuartzCore -framework Carbon"
   AM_CONDITIONAL(USE_QUARTZ, true)
   if test "x$quartz_relocation" = xyes; then
     AC_DEFINE([QUARTZ_RELOCATION], [1], [Use NSBundle functions to determine load paths for libraries, translations, etc.])
