diff --git a/gdk/quartz/gdkwindow-quartz.c b/gdk/quartz/gdkwindow-quartz.c
index e302c5c..c0e60d8 100644
--- a/gdk/quartz/gdkwindow-quartz.c	
+++ b/gdk/quartz/gdkwindow-quartz.c
@@ -134,7 +134,7 @@ gdk_quartz_window_get_nsview (GdkWindow *window)
   if (!GDK_IS_WINDOW_IMPL_QUARTZ(window->impl))
     return NULL;
 
-  return ((GdkWindowImplQuartz *)window->impl)->view;
+  return GDK_WINDOW_IMPL_QUARTZ(window->impl)->view;
 }
 
 NSWindow *
@@ -146,7 +146,7 @@ gdk_quartz_window_get_nswindow (GdkWindow *window)
   if (!GDK_IS_WINDOW_IMPL_QUARTZ(window->impl))
     return NULL;
 
-  return ((GdkWindowImplQuartz *)window->impl)->toplevel;
+  return GDK_WINDOW_IMPL_QUARTZ(window->impl)->toplevel;
 }
 
 static CGContextRef
@@ -422,7 +422,7 @@ _gdk_quartz_window_process_updates_recurse (GdkWindow *window,
           GdkWindowImplQuartz *toplevel_impl;
           NSWindow *nswindow;
 
-          toplevel_impl = (GdkWindowImplQuartz *)toplevel->impl;
+          toplevel_impl = GDK_WINDOW_IMPL_QUARTZ(toplevel->impl);
           nswindow = toplevel_impl->toplevel;
 #if MAC_OS_X_VERSION_MIN_REQUIRED < 101400
           /* In theory, we could skip the flush disabling, since we only
@@ -1581,7 +1581,7 @@ gdk_window_quartz_raise (GdkWindow *window)
         {
           GdkWindowImplQuartz *impl;
 
-          impl = (GdkWindowImplQuartz *)parent->impl;
+          impl = GDK_WINDOW_IMPL_QUARTZ(parent->impl);
 
           impl->sorted_children = g_list_remove (impl->sorted_children, window);
           impl->sorted_children = g_list_prepend (impl->sorted_children, window);
@@ -1612,7 +1612,7 @@ gdk_window_quartz_lower (GdkWindow *window)
         {
           GdkWindowImplQuartz *impl;
 
-          impl = (GdkWindowImplQuartz *)parent->impl;
+          impl = GDK_WINDOW_IMPL_QUARTZ(parent->impl);
 
           impl->sorted_children = g_list_remove (impl->sorted_children, window);
           impl->sorted_children = g_list_append (impl->sorted_children, window);
diff --git a/gdk/quartz/gdkevents-quartz.c b/gdk/quartz/gdkevents-quartz.c
index 0d492c1..3d73705 100644
--- a/gdk/quartz/gdkevents-quartz.c	
+++ b/gdk/quartz/gdkevents-quartz.c
@@ -387,7 +387,7 @@ get_window_point_from_screen_point (GdkWindow *window,
   NSPoint point;
   GdkQuartzNSWindow *nswindow;
 
-  nswindow = (GdkQuartzNSWindow*)(((GdkWindowImplQuartz *)window->impl)->toplevel);
+  nswindow = (GdkQuartzNSWindow*)(GDK_WINDOW_IMPL_QUARTZ(window->impl)->toplevel);
   point = [nswindow convertPointFromScreen:screen_point];
   *x = point.x;
   *y = window->height - point.y;
@@ -793,7 +793,7 @@ find_toplevel_for_mouse_event (NSEvent    *nsevent,
 
           toplevel = toplevel_under_pointer;
 
-          toplevel_impl = (GdkWindowImplQuartz *)toplevel->impl;
+          toplevel_impl = GDK_WINDOW_IMPL_QUARTZ(toplevel->impl);
 
           *x = x_tmp;
           *y = y_tmp;
@@ -1405,7 +1405,7 @@ test_resize (NSEvent *event, GdkWindow *toplevel, gint x, gint y)
   /* Resizing from the resize indicator only begins if an GDK_QUARTZ_LEFT_MOUSE_BUTTON
    * event is received in the resizing area.
    */
-  toplevel_impl = (GdkWindowImplQuartz *)toplevel->impl;
+  toplevel_impl = GDK_WINDOW_IMPL_QUARTZ(toplevel->impl);
   if ([toplevel_impl->toplevel showsResizeIndicator])
   if ([event type] == GDK_QUARTZ_LEFT_MOUSE_DOWN &&
       [toplevel_impl->toplevel showsResizeIndicator])
diff --git a/modules/input/imquartz.c b/modules/input/imquartz.c
index bdee6da..a9f9d7b 100644
--- a/modules/input/imquartz.c	
+++ b/modules/input/imquartz.c
@@ -362,7 +362,13 @@ quartz_set_cursor_location (GtkIMContext *context, GdkRectangle *area)
     return;
 
   nsview = gdk_quartz_window_get_nsview (qc->client_window);
+  if (!nsview)
+    return;
+
   win = (GdkWindow *)[ (GdkQuartzView*)nsview gdkWindow];
+  if (!win)
+    return;
+
   g_object_set_data (G_OBJECT (win), GIC_CURSOR_RECT, qc->cursor_rect);
 }
 
