diff --git a/gdk/quartz/gdkmonitor-quartz.c b/gdk/quartz/gdkmonitor-quartz.c
index f3f99a5..b90ba69 100644
--- a/gdk/quartz/gdkmonitor-quartz.c	
+++ b/gdk/quartz/gdkmonitor-quartz.c
@@ -32,8 +32,8 @@ gdk_quartz_monitor_get_workarea (GdkMonitor   *monitor,
 {
   GDK_QUARTZ_ALLOC_POOL;
 
-  NSArray *array = [NSScreen screens];
-  NSScreen* screen;
+  NSArray *array   = [NSScreen screens];
+  NSScreen* screen = NULL;
   for (id obj in array)
     {
       CGDirectDisplayID screen_id =
