diff --git a/Library/Homebrew/brew.sh b/Library/Homebrew/brew.sh
index cbe08846a..da398631a 100644
--- a/Library/Homebrew/brew.sh
+++ b/Library/Homebrew/brew.sh
@@ -425,7 +425,12 @@ then
   HOMEBREW_PRODUCT="Homebrew"
   HOMEBREW_SYSTEM="Macintosh"
   [[ "${HOMEBREW_PROCESSOR}" == "x86_64" ]] && HOMEBREW_PROCESSOR="Intel"
-  HOMEBREW_MACOS_VERSION="$(/usr/bin/sw_vers -productVersion)"
+  # Allow setting of a different deployment target
+  # CLT SDK *must* be installed
+  if [[ -z "${HOMEBREW_MACOS_VERSION}" ]]
+  then
+    HOMEBREW_MACOS_VERSION="$(/usr/bin/sw_vers -productVersion)"
+  fi
   # Don't change this from Mac OS X to match what macOS itself does in Safari on 10.12
   HOMEBREW_OS_USER_AGENT_VERSION="Mac OS X ${HOMEBREW_MACOS_VERSION}"
