diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4ff3401..01161f8 100644
--- a/CMakeLists.txt	
+++ b/CMakeLists.txt
@@ -147,6 +147,9 @@ elseif(${CMAKE_SYSTEM_NAME} MATCHES "FreeBSD")
   add_definitions(-DOS_FREEBSD)
 elseif(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
   add_definitions(-DOS_MACOSX)
+
+  set(CMAKE_MACOS_RPATH TRUE)
+  set(CMAKE_INSTALL_NAME_DIR "${CMAKE_INSTALL_PREFIX}/lib/")
 endif()
 
 function(transform_sources_list INPUT_FILE OUTPUT_FILE)
