diff --git a/gdk/quartz/gdkevents-quartz.c b/gdk/quartz/gdkevents-quartz.c
index f94fd978a0..ba799194c1 100644
--- a/gdk/quartz/gdkevents-quartz.c
+++ b/gdk/quartz/gdkevents-quartz.c
@@ -855,10 +855,27 @@ find_window_for_ns_event (NSEvent *nsevent,
       /* Only handle our own entered/exited events, not the ones for the
        * titlebar buttons.
        */
-      if ([view trackingRect] == [nsevent trackingNumber])
-        return toplevel;
+      if ([view trackingRect] == nsevent.trackingNumber)
+        {
+          return toplevel;
+        }
+      /* NSEVent returns a bogus trackingNumber and a null
+         trackingArea most of the time in macOS 13 so fallback on the
+         window being in the expected place.
+       */
+      else if (!nsevent.trackingArea)
+        {
+          int x, y;
+          GdkDisplay* display =  gdk_window_get_display (toplevel);
+          if (toplevel == find_toplevel_under_pointer (display, NSEvent.mouseLocation, &x, &y))
+            return toplevel;
+          else
+            return NULL;
+        }
       else
-        return NULL;
+        {
+          return NULL;
+        }

     case GDK_QUARTZ_KEY_DOWN:
     case GDK_QUARTZ_KEY_UP:
