--- gdk/quartz/gdkwindow-quartz.c 	2025-01-25 05:25:12
+++ gdk/quartz/gdkwindow-quartz.c	2025-06-05 11:18:49
@@ -1391,6 +1391,7 @@

       frame_rect = [impl->toplevel frameRectForContentRect:content_rect];
       [impl->toplevel setFrame:frame_rect display:YES];
+      [impl->view setFrame:frame_rect];
       impl->cairo_surface = gdk_quartz_ref_cairo_surface (window);
       cairo_surface_destroy (impl->cairo_surface); // Remove the extra reference
     }
