diff --git a/modules/meson.build b/modules/meson.build
index f7e9992dc0..ce90da33c0 100644
--- a/modules/meson.build
+++ b/modules/meson.build
@@ -86,7 +86,7 @@ foreach module : modules
   link = module.get('link', [])
   link_args = module.get('link-args', [])
 
-  library(name,
+  shared_module(name,
     srcs,
     include_directories: rootInclude,
     dependencies: modules_deps + [ deps ],
@@ -94,5 +94,7 @@ foreach module : modules
     link_args: link_args,
     install: true,
     install_dir: gimpplugindir / 'modules',
+    # Explicitly set .so suffix on macOS
+    name_suffix: host_machine.system() == 'darwin' ? 'so' : [],
   )
-endforeach
+endforeach
\ No newline at end of file
