diff --git a/app/main.c b/app/main.c
index e25edd14af..c5a5416414 100644
--- a/app/main.c
+++ b/app/main.c
@@ -380,6 +380,14 @@ gimp_macos_setenv (const char * progname)
           need_pythonhome = FALSE;
         }
       g_free (tmp);
+      /* Detect we were built in MacPorts for MacOS */
+      tmp = g_strdup_printf ("%s/Library/Frameworks/Python.framework", res_dir);
+      if (tmp && !stat (tmp, &sb) && S_ISDIR (sb.st_mode))
+        {
+          g_print ("GIMP was built with MacPorts\n");
+          need_pythonhome = FALSE;
+        }
+      g_free (tmp);

       path_len = strlen (g_getenv ("PATH") ? g_getenv ("PATH") : "") + strlen (app_dir) + 2;
       path = g_try_malloc (path_len);
diff --git a/meson.build b/meson.build
index 6f5a504344..7875983295 100644
--- a/meson.build
+++ b/meson.build
@@ -960,7 +960,7 @@ perl = find_program('perl5', 'perl', 'perl5.005', 'perl5.004', 'perl')

 python3_minver = '>=3.6'

-python = pythonmod.find_installation('python3')
+python = pythonmod.find_installation('@@PYTHON3_BIN@@')
 message('Found Python @0@'.format(python.language_version()))

 have_python = get_option('python')
