diff --git gtk/gtkdnd-quartz.c gtk/gtkdnd-quartz.c
index 755abcb..a3d0081 100644
--- gtk/gtkdnd-quartz.c	
+++ gtk/gtkdnd-quartz.c
@@ -1975,7 +1975,7 @@ _gtk_drag_source_handle_event (GtkWidget *widget,
   switch (event->type)
     {
     case GDK_DROP_FINISHED:
-      result = (gdk_drag_context_get_dest_window (context) != NULL) ? GTK_DRAG_RESULT_SUCCESS : GTK_DRAG_RESULT_NO_TARGET;
+      result = (context->action != NULL) ? GTK_DRAG_RESULT_SUCCESS : GTK_DRAG_RESULT_NO_TARGET;
       gtk_drag_drop_finished (info, result);
       break;
     default:
